presto-orc/src/main/java/com/facebook/presto/orc/writer/SegmentedSliceBlockBuilder.java

```diff
@@ -0,0 +1,416 @@
+ * Each segment has Slice(data is concatenated and stored in one slice)
+ * and offsets to capture the start offset and length. New slices are appended
+ * to the open segment. Once the segment is full the segment is
+ * finalized and appended to the closed segments. A new open segment is
+ * created for further appends.
+ */
+public class SegmentedSliceBlockBuilder
+        extends AbstractVariableWidthBlock
+        implements BlockBuilder
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(SegmentedSliceBlockBuilder.class).instanceSize();
+
+    private final DynamicSliceOutput openSliceOutput;
+
+    private int openSegment;
+    private int openSegmentOffset;
+    private int[][] offsets;
+    private Slice[] closedSlices;
+    private long closedSlicesRetainedSize;
+    private long closedSlicesSizeInBytes;
+
+    public SegmentedSliceBlockBuilder(int expectedEntries, int expectedBytes)
+    {
+        int initialSize = Math.max(SegmentHelper.INITIAL_SEGMENTS, SegmentHelper.segment(expectedEntries) + 1);
+        offsets = new int[initialSize][];
+        closedSlices = new Slice[initialSize];
+        offsets[0] = new int[SegmentHelper.SEGMENT_SIZE + 1];
+        openSliceOutput = new DynamicSliceOutput(expectedBytes);
+    }
+
+    public void reset()
+    {
+        openSliceOutput.reset();
+        Arrays.fill(closedSlices, null);
+        // Fill the first offset array with 0, and free up the rest of the offsets array.
+        Arrays.fill(offsets[0], 0);
+        Arrays.fill(offsets, 1, offsets.length, null);
+        openSegment = 0;
+        openSegmentOffset = 0;
+        closedSlicesRetainedSize = 0;
+        closedSlicesSizeInBytes = 0;
+    }
+
+    @Override
+    public int getPositionOffset(int position)
+    {
+        return getOffset(position);
+    }
+
+    @Override
+    public int getSliceLength(int position)
+    {
+        int offset = SegmentHelper.offset(position);
+        int segment = SegmentHelper.segment(position);
+        return offsets[segment][offset + 1] - offsets[segment][offset];
+    }
+
+    private Slice getSegmentRawSlice(int segment)
+    {
+        return segment == openSegment ? openSliceOutput.getUnderlyingSlice() : closedSlices[segment];
+    }
+
+    @Override
+    public Slice getRawSlice(int position)
+    {
+        return getSegmentRawSlice(SegmentHelper.segment(position));
+    }
+
+    @Override
+    public int getPositionCount()
+    {
+        return SegmentHelper.getPositions(openSegment, openSegmentOffset);
+    }
+
+    @Override
+    public long getSizeInBytes()
+    {
+        long arraysSizeInBytes = Integer.BYTES * (long) getPositionCount();
+        return openSliceOutput.size() + arraysSizeInBytes + closedSlicesSizeInBytes;
+    }
+
+    @Override
+    public long getRegionSizeInBytes(int positionOffset, int length)
+    {
+        throw new UnsupportedOperationException(getClass().getName());
```

Comment:
Usually we put explicit error messages like "getRegionSizeInBytes is not supported by ... "

Reply:
done.